### PR TITLE
Revamp the connector to latest ASB SDK 7.13.1

### PR DESCRIPTION
**Pull request by SanduDS**
# Description

The connector was updated to the Azure service messaging bus latest version - 7.13.1

Fixes # (https://github.com/wso2-enterprise/choreo/issues/17916)

One line release note: 
- ASB SDK update to the latest version 7.13.1

## Type of change

Please delete options that are not relevant.

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

All connector functions were tested with Azure service bus account and verified.

**Test Configuration**:
* Ballerina Version: Swan Lake 2201.3.1
* Operating System: Ubuntu
* Java SDK: 11

# Checklist:

### Security checks
 - [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? 
 - [X] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: asb-ballerina/Ballerina.toml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 [package]
-distribution = "2201.3.0"
+distribution = "2201.3.1"
 org = "ballerinax"
 name = "asb"
-version = "2.4.0"
+version = "2.5.0"
```
Comment (SkNuwanTissera) [inline asb-ballerina/Ballerina.toml:5]: Don't we need to release this as a major update `3.0.0`? 

WDYT? @LakshanSS @indikasampath2000 @abeykoon

File: asb-ballerina/messageReceiver.bal [hunk 2]
```diff
@@ -22,25 +22,36 @@ import ballerina/jballerina.java as java;
...
+    # + connectionString - Connection String of Azure service bus  
+    # + queueName - Queue name  
+    # + topicName - Topic name  
+    # + subscriptionName - Subscription name  
     # + receiveMode - Receive mode as PEEKLOCK or RECEIVEANDDELETE (default : PEEKLOCK)
-    public isolated function init(@display {label: "Connection String"} string connectionString, @display {label: "Entity Path"} string entityPath, 
-                                  @display {label: "Receive Mode"} string receiveMode = PEEKLOCK) returns error? {
+    # + return - Return Value Description
+    public isolated function init(@display {label: "Connection String"} string connectionString, 
+            @display {label: "Entity Path"} string? queueName = null,
```
Comment (SkNuwanTissera) [inline asb-ballerina/messageReceiver.bal:43]:
```suggestion
            @display {label: "Queue Name"} string? queueName = null, 
```
3 labels need to be changed

File: asb-ballerina/messageReceiver.bal [hunk 3]
```diff
@@ -22,25 +22,36 @@ import ballerina/jballerina.java as java;
...
+    # + queueName - Queue name  
+    # + topicName - Topic name  
+    # + subscriptionName - Subscription name  
     # + receiveMode - Receive mode as PEEKLOCK or RECEIVEANDDELETE (default : PEEKLOCK)
-    public isolated function init(@display {label: "Connection String"} string connectionString, @display {label: "Entity Path"} string entityPath, 
-                                  @display {label: "Receive Mode"} string receiveMode = PEEKLOCK) returns error? {
+    # + return - Return Value Description
+    public isolated function init(@display {label: "Connection String"} string connectionString, 
+            @display {label: "Entity Path"} string? queueName = null, 
+            @display {label: "Entity Path"} string? topicName = null,
```
Comment (SkNuwanTissera) [inline asb-ballerina/messageReceiver.bal:44]:
```suggestion
            @display {label: "Topic Name"} string? topicName = null, 
```

File: asb-ballerina/messageReceiver.bal [hunk 4]
```diff
@@ -22,25 +22,36 @@ import ballerina/jballerina.java as java;
...
+    # + topicName - Topic name  
+    # + subscriptionName - Subscription name  
     # + receiveMode - Receive mode as PEEKLOCK or RECEIVEANDDELETE (default : PEEKLOCK)
-    public isolated function init(@display {label: "Connection String"} string connectionString, @display {label: "Entity Path"} string entityPath, 
-                                  @display {label: "Receive Mode"} string receiveMode = PEEKLOCK) returns error? {
+    # + return - Return Value Description
+    public isolated function init(@display {label: "Connection String"} string connectionString, 
+            @display {label: "Entity Path"} string? queueName = null, 
+            @display {label: "Entity Path"} string? topicName = null, 
+            @display {label: "Entity Path"} string? subscriptionName = null,
```
Comment (SkNuwanTissera) [inline asb-ballerina/messageReceiver.bal:45]:
```suggestion
            @display {label: "Subscription Name"} string? subscriptionName = null, 
```

File: asb-ballerina/messageSender.bal [hunk 5]
```diff
@@ -22,21 +22,26 @@ import ballerina/jballerina.java as java;
...
     # Configure the connection string to have the [required permission](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-sas).
     # 
     # + connectionString - Connection String of Azure service bus
-    # + entityPath - Name or path of the entity (e.g : Queue name, Topic name)
+    # + queueName - Queue name
+    # + topicName - Topic name
     public isolated function init(@display {label: "Connection String"} string connectionString, 
-                                  @display {label: "Entity Path"} string entityPath) returns error? {
+                                  @display {label: "Queue Path"} string? queueName = null,
+                                  @display {label: "Queue Path"} string? topicName = null) returns error? {
```
Comment (SkNuwanTissera) [inline asb-ballerina/messageSender.bal:39]:
```suggestion
                                  @display {label: "Topic Name"} string? topicName = null) returns error? {
```

File: asb-ballerina/Ballerina.toml (hunk 1 above)
Comment (LakshanSS) [reply asb-ballerina/Ballerina.toml:5]: If we introduce any api changes or any major breaking changes, we can go to `3.0.0`.
If not, we can go with `2.5.0`

File: gradle.properties [hunk 6]
```diff
@@ -1,4 +1,4 @@
 org.gradle.caching=true
 group=org.ballerinax.azure.servicebus
-version=2.4.0
+version=2.5.0
 ballerinaLangVersion=2201.3.0-rc1
```
Comment (LakshanSS) [inline gradle.properties:4]:
```suggestion
ballerinaLangVersion=2201.3.1
```

File: asb-ballerina/Ballerina.toml (hunk 1 above)
Comment (LakshanSS) [reply asb-ballerina/Ballerina.toml:5]: Seems like there are major changes done. So lets go with `3.0.0`

File: asb-ballerina/messageSender.bal [hunk 7]
```diff
@@ -22,21 +22,26 @@ import ballerina/jballerina.java as java;
...
     # Initializes the connector. During initialization you can pass the [Shared Access Signature (SAS) authentication credentials](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-sas)
     # Create an [Azure account](https://docs.microsoft.com/en-us/learn/modules/create-an-azure-account/) and 
     # obtain tokens following [this guide](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-quickstart-portal#get-the-connection-string). 
     # Configure the connection string to have the [required permission](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-sas).
     # 
     # + connectionString - Connection String of Azure service bus
-    # + entityPath - Name or path of the entity (e.g : Queue name, Topic name)
+    # + queueName - Queue name
+    # + topicName - Topic name
     public isolated function init(@display {label: "Connection String"} string connectionString,
```
Comment (abeykoon) [inline asb-ballerina/messageSender.bal:37]: Lets define the constructor as 

init (string entityName, EntityType entityType) {..}

EntityType contains --> QUEUE, TOPIC

File: asb-ballerina/messageReceiver.bal [hunk 8]
```diff
@@ -22,25 +22,36 @@ import ballerina/jballerina.java as java;
...
-    # + entityPath - Name or path of the entity (e.g : Queue name, Subscription path)
+    #
+    # + connectionString - Connection String of Azure service bus  
+    # + queueName - Queue name  
+    # + topicName - Topic name  
+    # + subscriptionName - Subscription name  
     # + receiveMode - Receive mode as PEEKLOCK or RECEIVEANDDELETE (default : PEEKLOCK)
-    public isolated function init(@display {label: "Connection String"} string connectionString, @display {label: "Entity Path"} string entityPath, 
-                                  @display {label: "Receive Mode"} string receiveMode = PEEKLOCK) returns error? {
+    # + return - Return Value Description
```
Comment (abeykoon) [inline asb-ballerina/messageReceiver.bal:41]: Shall we move string receiveMode to an enum? Then the options are suggested to the user.

File: asb-ballerina/commons.bal [hunk 9]
```diff
@@ -46,12 +38,77 @@ public enum receiveModes {
...
+# This record holds the configuration details of a topic and its associated subscription in Azure Service Bus
+#
+# + topicName - A string field that holds the name of the topic  
+# + subscriptionName - A string field that holds the name of the subscription associated with the topic
+@display {label: "Topic/Subscriptions Configurations"}
+public  type TopicSubsConfig record {
+   @display {label: "Topic Name"}
+   string topicName;
+   @display {label: "Subscription Name"}
+   string subscriptionName;
```
Comment (abeykoon) [inline asb-ballerina/commons.bal:69]: The subscriptionName is mandatory thing to provide?

File: asb-ballerina/message.bal [hunk 10]
```diff
@@ -33,7 +33,7 @@
 @display {label: "Message"}
 public type Message record {|
     @display {label: "Body"}
-    string|xml|json|byte[] body;
+    string|int|byte[] body;
```
Comment (abeykoon) [inline asb-ballerina/message.bal:36]: Let's add a comment here stating that these are the basic AMQP message types. (point the reference)

File: asb-ballerina/message.bal [hunk 11]
```diff
@@ -59,6 +59,7 @@ public type Message record {|
     @display {label: "Lock Token"}
     readonly string lockToken?;
     ApplicationProperties applicationProperties?;
+    byte[] amqpBody?;
```
Comment (abeykoon) [inline asb-ballerina/message.bal:62]: Why do we need this separately? Isn't `string|int|byte[] body` enough?

File: asb-ballerina/commons.bal [hunk 12]
```diff
@@ -21,22 +21,14 @@ public const int DEFAULT_TIME_TO_LIVE = 300; // In seconds
 public const int DEFAULT_MAX_MESSAGE_COUNT = 1;
 public const int DEFAULT_SERVER_WAIT_TIME = 300; // In seconds
```
Comment (SkNuwanTissera) [inline asb-ballerina/commons.bal:22]:
```suggestion
public const TEXT = "text/plain";
public const JSON = "application/json";
public const XML = "application/xml";
public const BYTE_ARRAY = "application/octet-stream";
```

Minor thing; Now we can remove`string` from constants

File: asb-ballerina/commons.bal [hunk 13]
```diff
@@ -46,12 +38,77 @@ public enum receiveModes {
...
+# + receiveMode - This field holds the receive modes(RECEIVE_AND_DELETE/PEEK_LOCK) for the connection. The receive mode determines how messages are 
+#                 retrieved from the entity. The default value is PEEK_LOCK
+@display {label: "Receiver Connection Config"}
+public type ASBServiceReceiverConfig record {
+    @display {label: "Entity Type"}
+    EntityType entityType;
+    @display {label: "Entity Configuration"}
+    TopicSubsConfig|QueueConfig entityConfig;
+    @display {label: "ConnectionString"}
+    string connectionString;
```
Comment (abeykoon) [inline asb-ballerina/commons.bal:55]: I know fields in a record has no order. But for config panel rendering purposes what we make order as

    @display {label: "ConnectionString"}
    string connectionString;
    @display {label: "Entity Type"}
    EntityType entityType;
    @display {label: "Entity Configuration"}
    TopicSubsConfig|QueueConfig entityConfig;

However, do we need `EntityType ` as it is captured by entityConfig record's type?

File: asb-ballerina/commons.bal [hunk 14]
```diff
@@ -46,12 +38,77 @@ public enum receiveModes {
...
+   string subscriptionName; 
+};
+
+# This record holds the configuration details of a queue in Azure Service Bus
+#
+# + QueueName - A string field that holds the name of the queue
+@display {label: "Queue Configurations"}
+public  type QueueConfig record {
+   @display {label: "Queue Name"}
+   string QueueName;
```
Comment (abeykoon) [inline asb-ballerina/commons.bal:78]: QueueName -> queueName?

File: asb-ballerina/commons.bal [hunk 15]
```diff
@@ -21,22 +21,14 @@ public const int DEFAULT_TIME_TO_LIVE = 300; // In seconds
 public const int DEFAULT_MAX_MESSAGE_COUNT = 1;
 public const int DEFAULT_SERVER_WAIT_TIME = 300; // In seconds
 public const string DEFAULT_MESSAGE_LOCK_TOKEN = "00000000-0000-0000-0000-000000000000";
+public const string EMPTY_STRING = "";
 
 // Message content types
 public const string TEXT = "text/plain";
 public const string JSON = "application/json";
 public const string XML = "application/xml";
 public const string BYTE_ARRAY = "application/octet-stream";
```
Comment (SkNuwanTissera) [inline asb-ballerina/commons.bal:30]:
```suggestion
public const BYTE_ARRAY = "application/octet-stream";
```

Minor thing; Now we can remove `string` from constants

File: asb-ballerina/messageReceiver.bal [hunk 16]
```diff
@@ -21,26 +21,42 @@ import ballerina/jballerina.java as java;
...
+            TopicSubsConfig topicSubsConfig = check config.entityConfig.ensureType(TopicSubsConfig);
+            self.topicName = topicSubsConfig.topicName;
+            self.subscriptionName = topicSubsConfig.subscriptionName;
+            self.queueName = EMPTY_STRING;
+        }
+        
+        self.receiveMode = config.receiveMode;
         self.receiverHandle = check initMessageReceiver(java:fromString(self.connectionString), 
-        java:fromString(self.entityPath), java:fromString(self.receiveMode));
+        java:fromString(self.queueName),java:fromString(self.topicName), java:fromString(self.subscriptionName),
```
Comment (abeykoon) [inline asb-ballerina/messageReceiver.bal:58]: Why do we get `queueName ` , `topicName ` and `subscriptionName ` like this?

File: asb-ballerina/messageReceiver.bal [hunk 17]
```diff
@@ -104,7 +108,7 @@ public isolated client class MessageReceiver {
     @display {label: "Abandon Message"}
     isolated remote function abandon(@display {label: "Message"} Message message) returns Error? {
         if message?.lockToken.toString() != DEFAULT_MESSAGE_LOCK_TOKEN {
-            return abandon(self.receiverHandle, message?.lockToken.toString());
+            return abandon(self.receiverHandle, self, message?.lockToken.toString());
```
Comment (abeykoon) [inline asb-ballerina/messageReceiver.bal:111]: message?.lockToken --> is lockToken optional? Is that not a mandatory field in the message?

File: asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java [hunk 18]
```diff
@@ -18,340 +18,460 @@
...
+            }
+        } else if (!subscriptionName.isEmpty() && !topicName.isEmpty()) {
+            if (Objects.equals(receiveMode, RECEIVE_AND_DELETE)) {
+                this.receiver = receiverClientBuilder
+                        .receiveMode(ServiceBusReceiveMode.RECEIVE_AND_DELETE)
+                        .topicName(topicName)
+                        .subscriptionName(subscriptionName)
+                        .buildClient();
+
+            } else {
```
Comment (abeykoon) [inline asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java:112]: Better tell somewhere in bal docs, PEEK_LOCK mode is the default.

File: asb-ballerina/commons.bal (hunk 9 above)
Comment (SanduDS) [reply asb-ballerina/commons.bal:69]: Yes, It is for Receiver. Previous version , it was provided along with the entity path like this : topicName/Subscription/subscriptionName, However latest 7.13.1 SDK needs them as separate parameters.

File: asb-ballerina/messageReceiver.bal (hunk 16 above)
Comment (SanduDS) [reply asb-ballerina/messageReceiver.bal:58]: Here, these values are sent to ASB receiver constructor and its parameters are Java strings. (java.lang.String). Therefore, we take ballerina strings and converts them to java type string by using java:fromString. then they are passed using ballerina handles. AFAIK and I tested removing these java type conversion, it is not possible to initialize the java native receiver class. Please refer here https://ballerina.io/learn/java-interoperability-guide/ballerina-ffi/#deal-with-overloaded-constructors

File: asb-ballerina/messageReceiver.bal (hunk 17 above)
Comment (SanduDS) [reply asb-ballerina/messageReceiver.bal:111]: Yes it is mandatory to receive messages in PEEK_LOCK mode. We use same message record to provide message to ASB sender client, where the lock_token is not mandatory.

File: asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java (hunk 18 above)
Comment (SanduDS) [reply asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java:112]: https://github.com/SanduDS/module-ballerinax-azure-service-bus/blob/cddb330359d18591b09b5c4f9c619e455840dee8/asb-ballerina/messageReceiver.bal#L40

already added here

File: asb-ballerina/Module.md [hunk 19]
```diff
@@ -18,7 +18,7 @@ applies [SAS authentication](https://docs.microsoft.com/en-us/azure/service-bus-
 supports SAS authentication as well.
 
 This module
-supports [Service Bus SDK 3.5.1 version](https://docs.microsoft.com/en-us/java/api/overview/azure/servicebus/client?view=azure-java-stable&preserve-view=true)
+supports [Service Bus SDK 7.13.1 version](https://docs.microsoft.com/en-us/java/api/overview/azure/servicebus/client?view=azure-java-stable&preserve-view=true)
```
Comment (RolandHewage) [inline asb-ballerina/Module.md:21]: Let's update the link to the latest SDK as well.
https://learn.microsoft.com/en-us/java/api/overview/azure/messaging-servicebus-readme?view=azure-java-stable

File: asb-ballerina/Module.md [hunk 20]
```diff
@@ -18,7 +18,7 @@ applies [SAS authentication](https://docs.microsoft.com/en-us/azure/service-bus-
...
 . The source code on GitHub is
 located [here](https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/servicebus/microsoft-azure-servicebus). The
```
Comment (RolandHewage) [inline asb-ballerina/Module.md:23]: Let's update this pointing to the latest SDK as well
https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/servicebus/azure-messaging-servicebus

File: asb-ballerina/Module.md [hunk 21]
```diff
@@ -61,10 +60,10 @@ Before using this connector in your Ballerina application, complete the followin
 
         2. [Get the connection string](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-quickstart-portal#get-the-connection-string)
 
-        3. [Create a topic in the Azure portal & get Entity Path](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-quickstart-topics-subscriptions-portal#create-a-topic-using-the-azure-portal)
+        3. [Create a topic in the Azure portal & get topic name](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-quickstart-topics-subscriptions-portal#create-a-topic-using-the-azure-portal)
            . It's in the format ‘topicName‘.
 
-        4. [Create a subscription in the Azure portal & get Entity Path](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-quickstart-topics-subscriptions-portal#create-subscriptions-to-the-topic)
+        4. [Create a subscription in the Azure portal & get its name](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-quickstart-topics-subscriptions-portal#create-subscriptions-to-the-topic)
```
Comment (RolandHewage) [inline asb-ballerina/Module.md:66]: suggestion: it's

File: asb-ballerina/commons.bal [hunk 22]
```diff
@@ -46,12 +38,77 @@ public enum receiveModes {
...
+    TOPIC = "topic"
+}
+
+//Message receiver modes
+public enum ReceiveMode {
+    @display {label: "RECEIVEANDDELETE"}
+    RECEIVE_AND_DELETE = "RECEIVEANDDELETE",
+    @display {label: "PEEKLOCK"}
+    PEEK_LOCK = "PEEKLOCK"
+}
```
Comment (RolandHewage) [inline asb-ballerina/commons.bal:111]: We may use the same name for the enums. But the existing one is correct too.  
```
// Message receive modes
@display {label: "Receive Modes"}
public enum receiveModes {
    @display {label: "PEEKLOCK"}
    PEEKLOCK,
    @display {label: "RECEIVEANDDELETE"}
    RECEIVEANDDELETE
}
```

File: asb-ballerina/messageReceiver.bal [hunk 23]
```diff
@@ -21,26 +21,42 @@ import ballerina/jballerina.java as java;
...
+    # + queueName - Queue name  
+    # + topicName - Topic name  
+    # + subscriptionName - Subscription name  
     # + receiveMode - Receive mode as PEEKLOCK or RECEIVEANDDELETE (default : PEEKLOCK)
-    public isolated function init(@display {label: "Connection String"} string connectionString, @display {label: "Entity Path"} string entityPath, 
-                                  @display {label: "Receive Mode"} string receiveMode = PEEKLOCK) returns error? {
-        self.connectionString = connectionString;
-        self.entityPath = entityPath;
-        self.receiveMode = receiveMode;
+    # + return - Return Value Description
```
Comment (RolandHewage) [inline asb-ballerina/messageReceiver.bal:41]: These doc params are not passed in the init method. Add them above private variable declarations if necessary.

File: asb-ballerina/messageSender.bal [hunk 24]
```diff
@@ -21,22 +21,24 @@ import ballerina/jballerina.java as java;
...
     # 
-    # + connectionString - Connection String of Azure service bus
-    # + entityPath - Name or path of the entity (e.g : Queue name, Topic name)
-    public isolated function init(@display {label: "Connection String"} string connectionString, 
-                                  @display {label: "Entity Path"} string entityPath) returns error? {
-        self.connectionString = connectionString;
-        self.entityPath = entityPath;
-        self.senderHandle = check initMessageSender(java:fromString(self.connectionString), java:fromString(self.entityPath));
+    # + connectionString - ConnectionString of Azure service bus
+    # + topicOrQueueName - Topic or Queue name
```
Comment (RolandHewage) [inline asb-ballerina/messageSender.bal:35]: Let's add the doc comment for the `ASBServiceSenderConfig` here

File: asb-ballerina/Module.md (hunk 21 above)
Comment (SanduDS) [reply asb-ballerina/Module.md:66]: it's = it is ? or its? I guess "its"

File: asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java [hunk 25]
```diff
@@ -18,340 +18,437 @@
...
-        values[12] = StringUtils.fromString(receivedMessage.getLockToken().toString());
-        BMap<BString, Object> applicationProperties =
-                ValueCreator.createRecordValue(ModuleUtils.getModule(), ASBConstants.APPLICATION_PROPERTIES);
-        Object[] propValues = new Object[1];
-        propValues[0] = ASBUtils.toBMap(receivedMessage.getProperties());
-        values[13] = ValueCreator.createRecordValue(applicationProperties, propValues);
-        BMap<BString, Object> messageRecord =
-                ValueCreator.createRecordValue(ModuleUtils.getModule(), ASBConstants.MESSAGE_RECORD);
-        return ValueCreator.createRecordValue(messageRecord, values);
+    public Object convertAMQPToJava(Object amqpValue) {
```
Comment (RolandHewage) [inline asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java:160]: Let's add a doc comment here

**Review by RolandHewage — COMMENTED** [review]
Added few comments. Rest seems fine.

**Review by abeykoon — APPROVED** [review]

File: asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java [hunk 26]
```diff
@@ -18,340 +18,460 @@
...
+            options.setDeadLetterReason(ASBUtils.valueToEmptyOrToString(deadLetterReason));
+            receiver.deadLetter(message, options);
             if (log.isDebugEnabled()) {
                 log.debug("\tDone dead-lettering a message using its lock token from \n" + receiver.getEntityPath());
             }
             return null;
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            return null;
         } catch (ServiceBusException e) {
```
[the remark this answers is not in the document]
Comment (SanduDS) [reply asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java:396]: As we discussed we will handle  service bus exception. however. we might get exception according to the ampqp message and its body. so we had better handle like this.

[the remark this answers is not in the document]
Comment (SanduDS) [reply asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java:396]: As we discussed we will handle  service bus exception. however. we might get exception according to the amqp message and its body. so we had better handle like this.

File: asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java [hunk 27]
```diff
@@ -18,340 +18,460 @@
...
-            receiver.complete(UUID.fromString(lockToken.toString()));
-            return null;
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
             return null;
         } catch (ServiceBusException e) {
-            return ASBUtils.returnErrorValue("Exception while completing message" + e.getMessage());
+            return ASBUtils.returnErrorValue("ServiceBusException while completing message " + e.getMessage());
+        } catch (Exception e) {
+            return ASBUtils.returnErrorValue("Exception while completing message " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (SanduDS) [reply asb-native/src/main/java/org/ballerinax/asb/receiver/MessageReceiver.java:343]: Here we create an error message to be passed to ballerina side. as we discussed we added logs with e

File: asb-native/src/main/java/org/ballerinax/asb/sender/MessageSender.java [hunk 28]
```diff
@@ -120,35 +129,54 @@ public Object send(Object body, Object contentType, Object messageId,
...
+                    continue;
+                }
+                // The batch is full, so we create a new batch and send the batch.
+                sender.sendMessages(currentBatch);
+                currentBatch = sender.createMessageBatch();
+
+                // Add that message that we couldn't before.
+                if (!currentBatch.tryAddMessage(message)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Message is too large for an empty batch. Skipping. Max size: "
```
[the remark this answers is not in the document]
Comment (SanduDS) [reply asb-native/src/main/java/org/ballerinax/asb/sender/MessageSender.java:174]: According to the logic here and as I explained, we sent all the message using multiple batches and without worrying the size. this is a debug log for sending messages.

**Pull request merged**